### PR TITLE
Handle custom operations with overlapping names in QPY

### LINKED ISSUE #8941: qpy mishandles the `.definition` of custom gates with the same `.name` but different params
### Environment

- **Qiskit Terra version**: 0.22.0
- **Python version**: 3.8.13
- **Operating system**: Ubuntu 22.04.1

### What is happening?

When a circuit contains multiple instances of a custom gate (with the same `.name`), qpy serialization+deserialization will silently replace the `.definition` of each with that of the first (resulting in incorrect definitions for parameterized gates).

### How can we reproduce the issue?

```
import io
import qiskit.qpy
 
class MyParamGate(qiskit.circuit.Gate):
    def __init__(self, phi):
        super().__init__("my_gate", 1, [phi])

    def _define(self):
        qc = qiskit.QuantumCircuit(1)
        qc.rx(self.params[0], 0)
        self.definition = qc

qc = qiskit.QuantumCircuit(1)
qc.append(MyParamGate(1.1), [0])
qc.append(MyParamGate(1.2), [0])

print("Before serialization:")
print(qc)
print("with gate .definitions:")
for inst in qc:
    print(inst.operation.definition)


buf = io.BytesIO()
qiskit.qpy.dump(qc, buf)
qc = qiskit.qpy.load(io.BytesIO(buf.getvalue()))[0]

print(f"\nAfter serialization:")
print(qc)
print("with gate .definitions:")
for inst in qc:
    print(inst.operation.definition)
```
prints:
```
Before serialization:
   ┌──────────────┐┌──────────────┐
q: ┤ My_gate(1.1) ├┤ My_gate(1.2) ├
   └──────────────┘└──────────────┘
with gate .definitions:
   ┌─────────┐
q: ┤ Rx(1.1) ├
   └─────────┘
   ┌─────────┐
q: ┤ Rx(1.2) ├
   └─────────┘

After serialization:
   ┌──────────────┐┌──────────────┐
q: ┤ my_gate(1.1) ├┤ my_gate(1.2) ├
   └──────────────┘└──────────────┘
with gate .definitions:
   ┌─────────┐
q: ┤ Rx(1.1) ├
   └─────────┘
   ┌─────────┐
q: ┤ Rx(1.1) ├
   └─────────┘
```
where after serialization both gates' `.params` are correct, but the second gate's `.definition` has been replaced with that of the first

### What should happen?

the final line should read `Rx(1.2)` instead of `Rx(1.1)` (or otherwise `qpy.dump` should throw an error)

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit fixes a longstanding bug and limitation in QPY around the use of custom operations. When QPY encounters a non-standard (meaning not in Qiskit itself, or a direct instance of Gate or Instruction) instruction it is not able to represent the exact instance perfectly in that payload. This is because it would require more code that represents those classes or data not contained inside Qiskit itself, which would be an opportunity for arbitrary code execution which is something QPY was designed to prevent (as it's a limitation with pickle serialization). So to serialize these objects in a circuit, QPY stores as much data as it can extract from the data model from the Instruction and Gate classes and builds a table of custom instructions that will be recreated as Instruction or Gate instances on deserialization. In all previous QPY format versions that table was keyed on the .name attribute of the custom instructions in the circuit, the thinking being the name is a unique op code during circuit compilation and if there are multiple circuit elements with the same name they should be the same object. With this assumption it enabled the QPY payloads generated with repeated custom instructions to be smaller and more efficient because we don't need to store potentially repeated data. While that assumption is true during most of compilation it was ignoring that for a bare quantum circuit (or during the initial stages of compilation) this isn't necessarily true and there are compiler passes that canonicalize custom operations prior to that being true. To make it worse this assumption was causing conflicts in the QPY payload and cause an inaccurate reproduction of the original circuit when deserializing a QPY payload in many cases when custom gates were used.

This commit fixes this limitation by introducing a new QPY format version 11, which changes the key used in the custom instruction table to instead of being just the name to `{name}_{uuid}` where each instance of a custom instruction has a different uuid value. This means that there are never any overlapping definitions in the table and we don't have to worry about the case where two custom instructions have the same name.

### Details and comments

Fixes #8941

### CODE REVIEW

**Comment by mtreinish** [comment]
I wrote this with the intent to combine it with #11644 so there is a version kwarg added to some of the write functions which is not set anywhere externally. Depending on whether this or #11644 merges first we'll want to update the other to ensure that when `qpy.dump(version=10)` is called it connects to the `version` kwarg on `qiskit.qpy.binary_io.circuits.write_circuit()`.

**Review by ElePT — COMMENTED** [review]
This looks good to me, it's a bit of a shame that I didn't think of generalizing the fix in #10809 or at least stripping the trailing `_{uuid}` (version 10 outputs beautiful gate names like `ucrz_dg_0d501712-5682-4a9c-99ea-07453a0a3fe1(-π/2)`), but it's a lot better now. I only have two super minor comments. I can approve once the changes from https://github.com/Qiskit/qiskit/pull/11644 are merged.

File: qiskit/qpy/binary_io/circuits.py
Comment (ElePT) [inline qiskit/qpy/binary_io/circuits.py]: Maybe the order of the if/elif could be reverted? Not that it would change anything in the output, because the uuid is not outputted anymore, but it feels nicer to know that all the uuids are in the same format if version>=11.

File: test/python/qpy/test_circuit_load_from_qpy.py
Comment (ElePT) [inline test/python/qpy/test_circuit_load_from_qpy.py:149]: Wouldn't it make more sense to have this test be part of `TestLoadFromQPY`(in `test/python/circuit/test_circuit_load_from_qpy.py`) rather than `TestLayout`? Could this be an opportunity to merge both test files under `test/python/qpy`?

Comment (mtreinish) [reply test/python/qpy/test_circuit_load_from_qpy.py:149]: It does, but for this PR I just left it as is. I'll push a follow up PR after this merges that unifies the tests into a single module under `tests/python/qpy`

File: qiskit/qpy/binary_io/circuits.py
Comment (mtreinish) [reply qiskit/qpy/binary_io/circuits.py]: Oh hah, I did it this way on purpose to preserve the exact behavior because I didn't view the ucr* gates as being custom gates and didn't want to overreach in this PR and change more in the format revision than was necessary.  But you're right we can just subsume the custom logic for the ucr* gates in version 11 and have it all be consistent. I'll reverse the logic here

Comment (mtreinish) [reply qiskit/qpy/binary_io/circuits.py]: Done in: https://github.com/Qiskit/qiskit/pull/11646/commits/c88a9a2784eb511e571eae00bd1e81541fcd4ecd

**Review by ElePT — APPROVED** [review]
LGTM!

**Comment by jakelishman** [comment]
I disembarked this so that #10998 could go first (because it'll have conflicts with other things), but that's also just joined the queue, so I can put this back in now.

**Pull request merged**
